### ISSUE TITLE
IoT Edge modules disconnection after some days of deployment

### ISSUE THREAD

**Issue by amunozh** [Azure/iotedge#4229]
## Expected Behavior
I expect that all the modules run continuously without any interrumption. 

## Current Behavior
After some time running some IoT Edge modules disconnect from the edgeHub. The following message appears in edge Hub module:

 [WRN] - No subscription for receiving messages found for 4f3ea2243c2b4abeade19aba3ebafff0/BadgeTwinHandler\n

In this case the BadgeTwinHandler module is disconnected, but it could happen to any of the modules in the deployment. 

## Steps to Reproduce
Provide a detailed set of steps to reproduce the bug.
1. Deploy the modules using the azure portal
2. Wait for some days (In this case passed around 2 weeks). In this period the device could experiment 

Note: This behavior cannot be reproduced easily and it is not happening in all the devices. But it has happend multiple times. It was more frequent when we were using previous edge agent and edge hub container versions. 

### Output of `iotedge check`
I cannot obtain in this moment the "iotedge check" logs because the device is already in a production scenario.

### Device Information
* Host OS [e.g. Ubuntu 16.04, Ubuntu 18.04, Windows IoT Core]: Raspbian Stretch
* Architecture [e.g. amd64, arm32, arm64]:  arm32v7
* Container OS [e.g. Linux containers, Windows containers]:  Linux Containers

### Runtime Versions
* aziot-edged [run `iotedge version`]: 1.0.9.5
* Edge Agent [image tag (e.g. 1.0.0)]: 1.0.10
* Edge Hub [image tag (e.g. 1.0.0)]: 1.0.10
* Docker/Moby [run `docker version`]: 3.0.13+azure

Note: when using Windows containers on Windows, run `docker -H npipe:////./pipe/iotedge_moby_engine version` instead

## Logs

<details>
<summary>aziot-edged logs</summary>

```
I cannot obtain in this moment the "iotedge check" logs because the device is already in a production scenario.
```
</details>

<details>
<summary>edge-hub logs</summary>

[edgeHub_logs.log](https://github.com/Azure/iotedge/files/5816078/edgeHub_logs.log)

</details>

<details>
<summary>Contactmanager logs</summary>

[ContactManager_logs.log](https://github.com/Azure/iotedge/files/5816048/ContactManager_logs.log)

</details>

<details>
<summary>badgeTwinHandler logs</summary>

[BadgeTwinHandler_logs.log](https://github.com/Azure/iotedge/files/5816041/BadgeTwinHandler_logs.log)

</details>

## Additional Information
In the first versions of edge agent and edge hub containers (1.0) this problem occurs frequelty (at least each 2 days), I update the containers hoping that the problem would dissapear completely, but it's still happening.  I think it occurs when the module is in the SAS reneweal token procedure, for some reason it cannot reconnect to the edge hub. The big problem is that the module never reconnects, so all the messagges remains into the edge hub.
To resolve the problem I restart the two modules that appeared disconnected. Once they restart they connect correctly to the edge Hub.

**Comment by lfitchett** [comment]
Are you using the IoT SDK? If so, what version?

**Comment by amunozh** [comment]
Yes, I'm using the Python SDK `azure-iot-device` version 2.4.0

**Comment by MattCost** [comment]
We are experiencing the same issue with some python based modules.
What version information would you need from us?

**Comment by MattCosturos** [comment]
I have found that Edge Hub will periodically restart, and when it restarts we have 1 module that will no longer receive messages.

 "No subscription for receiving messages found for DeviceName/ModuleName"
This repeats until we restart the ModuleName.

What logging info would be helpful for debugging this issue?

**Comment by massand** [comment]
There was a fix related to Python SDK reconnection issues with edgeHub. I see the fix is in edgeHub 1.2 container only at the moment and the 1.2.0-rc4 preview image should be compatible with Edge LTS 1.1 package.

Here is the fix -
https://github.com/Azure/iotedge/commit/7e2d191cb8b4cd45bcec7653c523b8f8ab5e707f#diff-4a9aa713ef0a6c8573feedd1207f15f495d90209826ca233a4c35a51b50e783cR46

Please set the problematic Edge device's edgeHub module image to "mcr.microsoft.com/azureiotedge-agent:1.2.0-rc4" in the Azure portal > Edge device's runtime settings > EdgeHub container image setting.

**Comment by MattCosturos** [comment]
Do you mean "mcr.microsoft.com/azureiotedge-hub:1.2.0-rc4" for the edgeHub image?

**Comment by amunozh** [comment]
> There was a fix related to Python SDK reconnection issues with edgeHub. I see the fix is in edgeHub 1.2 container only at the moment and the 1.2.0-rc4 preview image should be compatible with Edge LTS 1.1 package.
> 
> Here is the fix -
> [7e2d191#diff-4a9aa713ef0a6c8573feedd1207f15f495d90209826ca233a4c35a51b50e783cR46](https://github.com/Azure/iotedge/commit/7e2d191cb8b4cd45bcec7653c523b8f8ab5e707f#diff-4a9aa713ef0a6c8573feedd1207f15f495d90209826ca233a4c35a51b50e783cR46)
> 
> Please set the problematic Edge device's edgeHub module image to ~"mcr.microsoft.com/azureiotedge-agent:1.2.0-rc4"~ "mcr.microsoft.com/azureiotedge-hub:1.2.0-rc4" in the Azure portal > Edge device's runtime settings > EdgeHub container image setting.



Was this problem present also on 1.0.10 Edge Hub version? Because I think the problem is related with the reconnection with the cloud when the network is not stable.

**Comment by massand** [comment]
The latest 1.2.0 edgeHub image has been released. Since we haven't heard back if this issue can be reproduced in 1.2.0-rc4 or later edgeHub images, we are closing this issue for now.

**Comment by amunozh** [comment]
I'm reopening this issue because the problem is presenting again. Maybe it was solved on version 1.2 but my devices are deployed and I'm not able to update to version 1.2 because I'd need to change the iotedged daemon (am I mistaken?). 

Does anyone know why is it happening? And how to solve or workaround it? 

Thank you.

Here in below the logs, as you can see without any advice some modules get disconnected from edgeHub:

```

<6> 2021-05-12 13:23:44.852 +00:00 [INF] - ClientAuthenticated, ##/DFUModule, 31cb5eb0
<6> 2021-05-12 13:23:44.986 +00:00 [INF] - New device connection for device ##/BadgeTwinHandler
<6> 2021-05-12 13:23:44.987 +00:00 [INF] - Client ##/BadgeTwinHandler connected to edgeHub, processing existing subscriptions.
<6> 2021-05-12 13:23:44.988 +00:00 [INF] - New device connection for device ##/DFUModule
<6> 2021-05-12 13:23:44.989 +00:00 [INF] - New device connection for device ##/UARTHandler
<6> 2021-05-12 13:23:44.989 +00:00 [INF] - Client ##/UARTHandler connected to edgeHub, processing existing subscriptions.
<6> 2021-05-12 13:23:44.989 +00:00 [INF] - Client ##/DFUModule connected to edgeHub, processing existing subscriptions.
<6> 2021-05-12 13:23:44.989 +00:00 [INF] - New device connection for device ##/ContactManager
<6> 2021-05-12 13:23:44.991 +00:00 [INF] - Client ##/ContactManager connected to edgeHub, processing existing subscriptions.
<6> 2021-05-12 13:23:45.001 +00:00 [INF] - New device connection for device ##/FOTAHandler
<6> 2021-05-12 13:23:45.001 +00:00 [INF] - Client ##/FOTAHandler connected to edgeHub, processing existing subscriptions.
<6> 2021-05-12 13:23:45.015 +00:00 [INF] - Attempting to connect to IoT Hub for client ##/FOTAHandler via AMQP over WebSockets...
<6> 2021-05-12 13:23:45.052 +00:00 [INF] - Bind device proxy for device ##/FOTAHandler
<6> 2021-05-12 13:23:45.052 +00:00 [INF] - Bind device proxy for device ##/UARTHandler
<6> 2021-05-12 13:23:45.053 +00:00 [INF] - Bind device proxy for device ##/ContactManager
<6> 2021-05-12 13:23:45.052 +00:00 [INF] - Bind device proxy for device ##/BadgeTwinHandler
<6> 2021-05-12 13:23:45.054 +00:00 [INF] - Bind device proxy for device ##/DFUModule
<6> 2021-05-12 13:23:45.056 +00:00 [INF] - Binding message channel for device Id ##/UARTHandler
<6> 2021-05-12 13:23:45.057 +00:00 [INF] - Binding message channel for device Id ##/DFUModule
<6> 2021-05-12 13:23:45.056 +00:00 [INF] - Binding message channel for device Id ##/FOTAHandler
<6> 2021-05-12 13:23:45.057 +00:00 [INF] - Binding message channel for device Id ##/ContactManager
<6> 2021-05-12 13:23:45.056 +00:00 [INF] - Binding message channel for device Id ##/BadgeTwinHandler
<6> 2021-05-12 13:23:45.059 +00:00 [INF] - Attempting to connect to IoT Hub for client ##/UARTHandler via AMQP over WebSockets...
<6> 2021-05-12 13:23:45.063 +00:00 [INF] - Attempting to connect to IoT Hub for client ##/DFUModule via AMQP over WebSockets...
<6> 2021-05-12 13:23:45.065 +00:00 [INF] - Attempting to connect to IoT Hub for client ##/ContactManager via AMQP over WebSockets...
<6> 2021-05-12 13:23:45.261 +00:00 [INF] - Processing subscriptions ModuleMessages for client ##/DFUModule.
<6> 2021-05-12 13:23:45.262 +00:00 [INF] - Processing pending subscriptions for ##/DFUModule
<6> 2021-05-12 13:23:45.303 +00:00 [INF] - Attempting to connect to IoT Hub for client ##/BadgeTwinHandler via AMQP over WebSockets...
<6> 2021-05-12 13:23:45.352 +00:00 [INF] - Set subscriptions from session state for ##/DFUModule
<6> 2021-05-12 13:23:45.614 +00:00 [INF] - Cloud connection for ##/FOTAHandler is True
<6> 2021-05-12 13:23:45.615 +00:00 [INF] - Connection status for ##/FOTAHandler changed to ConnectionEstablished
<6> 2021-05-12 13:23:45.615 +00:00 [INF] - Client ##/FOTAHandler connected to cloud, processing existing subscriptions.
<6> 2021-05-12 13:23:45.615 +00:00 [INF] - Skipping ##/FOTAHandler for subscription processing, as it is currently being processed.
<6> 2021-05-12 13:23:45.616 +00:00 [INF] - Created cloud proxy for client ##/FOTAHandler via AMQP over WebSockets, with client operation timeout 20 seconds.
<6> 2021-05-12 13:23:45.616 +00:00 [INF] - Initialized cloud proxy a94fe53b-54d3-4203-b255-c287b5a39628 for ##/FOTAHandler
<6> 2021-05-12 13:23:45.617 +00:00 [INF] - Created cloud connection for client ##/FOTAHandler
<6> 2021-05-12 13:23:45.670 +00:00 [INF] - Cloud connection for ##/ContactManager is True
<6> 2021-05-12 13:23:45.671 +00:00 [INF] - Connection status for ##/ContactManager changed to ConnectionEstablished
<6> 2021-05-12 13:23:45.671 +00:00 [INF] - Client ##/ContactManager connected to cloud, processing existing subscriptions.
<6> 2021-05-12 13:23:45.671 +00:00 [INF] - Skipping ##/ContactManager for subscription processing, as it is currently being processed.
<6> 2021-05-12 13:23:45.672 +00:00 [INF] - Created cloud proxy for client ##/ContactManager via AMQP over WebSockets, with client operation timeout 20 seconds.
<6> 2021-05-12 13:23:45.672 +00:00 [INF] - Initialized cloud proxy 9e1ede78-5054-454d-acfa-08052ae6e639 for ##/ContactManager
<6> 2021-05-12 13:23:45.672 +00:00 [INF] - Created cloud connection for client ##/ContactManager
<6> 2021-05-12 13:23:45.681 +00:00 [INF] - Cloud connection for ##/DFUModule is True
<6> 2021-05-12 13:23:45.681 +00:00 [INF] - Connection status for ##/DFUModule changed to ConnectionEstablished
<6> 2021-05-12 13:23:45.681 +00:00 [INF] - Client ##/DFUModule connected to cloud, processing existing subscriptions.
<6> 2021-05-12 13:23:45.681 +00:00 [INF] - Skipping ##/DFUModule for subscription processing, as it is currently being processed.
<6> 2021-05-12 13:23:45.682 +00:00 [INF] - Created cloud proxy for client ##/DFUModule via AMQP over WebSockets, with client operation timeout 20 seconds.
<6> 2021-05-12 13:23:45.682 +00:00 [INF] - Initialized cloud proxy c8dfe436-88f7-4074-b90c-c796f2f0ad19 for ##/DFUModule
<6> 2021-05-12 13:23:45.682 +00:00 [INF] - Created cloud connection for client ##/DFUModule
<6> 2021-05-12 13:23:45.687 +00:00 [INF] - Cloud connection for ##/UARTHandler is True
<6> 2021-05-12 13:23:45.687 +00:00 [INF] - Connection status for ##/UARTHandler changed to ConnectionEstablished
<6> 2021-05-12 13:23:45.687 +00:00 [INF] - Client ##/UARTHandler connected to cloud, processing existing subscriptions.
<6> 2021-05-12 13:23:45.687 +00:00 [INF] - Cloud connection for ##/BadgeTwinHandler is True
<6> 2021-05-12 13:23:45.687 +00:00 [INF] - Skipping ##/UARTHandler for subscription processing, as it is currently being processed.
<6> 2021-05-12 13:23:45.688 +00:00 [INF] - Connection status for ##/BadgeTwinHandler changed to ConnectionEstablished
<6> 2021-05-12 13:23:45.688 +00:00 [INF] - Created cloud proxy for client ##/UARTHandler via AMQP over WebSockets, with client operation timeout 20 seconds.
<6> 2021-05-12 13:23:45.688 +00:00 [INF] - Client ##/BadgeTwinHandler connected to cloud, processing existing subscriptions.
<6> 2021-05-12 13:23:45.688 +00:00 [INF] - Skipping ##/BadgeTwinHandler for subscription processing, as it is currently being processed.
<6> 2021-05-12 13:23:45.688 +00:00 [INF] - Initialized cloud proxy 09e5fb29-b669-41c9-92ee-cde06fea4777 for ##/UARTHandler
<6> 2021-05-12 13:23:45.688 +00:00 [INF] - Created cloud proxy for client ##/BadgeTwinHandler via AMQP over WebSockets, with client operation timeout 20 seconds.
<6> 2021-05-12 13:23:45.688 +00:00 [INF] - Created cloud connection for client ##/UARTHandler
<6> 2021-05-12 13:23:45.688 +00:00 [INF] - Initialized cloud proxy a8f6da4f-32ae-4d2e-8832-bca4c849d683 for ##/BadgeTwinHandler
<6> 2021-05-12 13:23:45.688 +00:00 [INF] - Created cloud connection for client ##/BadgeTwinHandler
<6> 2021-05-12 13:23:48.046 +00:00 [INF] - Updated reported properties for ##/$edgeHub
<4> 2021-05-12 13:23:50.263 +00:00 [WRN] - No subscription for receiving messages found for ##/ContactManager
<4> 2021-05-12 13:23:50.352 +00:00 [WRN] - No subscription for receiving messages found for ##/BadgeTwinHandler
<4> 2021-05-12 13:23:50.416 +00:00 [WRN] - No subscription for receiving messages found for ##/BadgeTwinHandler
<4> 2021-05-12 13:23:51.318 +00:00 [WRN] - No subscription for receiving messages found for ##/ContactManager
<4> 2021-05-12 13:23:51.354 +00:00 [WRN] - No subscription for receiving messages found for ##/BadgeTwinHandler
<4> 2021-05-12 13:23:51.414 +00:00 [WRN] - No subscription for receiving messages found for ##/BadgeTwinHandler
<4> 2021-05-12 13:23:53.277 +00:00 [WRN] - No subscription for receiving messages found for ##/BadgeTwinHandler
<4> 2021-05-12 13:23:53.413 +00:00 [WRN] - No subscription for receiving messages found for ##/ContactManager
<6> 2021-05-12 13:23:53.446 +00:00 [INF] - Done syncing reported properties for ##/$edgeHub
<4> 2021-05-12 13:23:53.496 +00:00 [WRN] - No subscription for receiving messages found for ##/BadgeTwinHandler
<4> 2021-05-12 13:23:57.211 +00:00 [WRN] - No subscription for receiving messages found for ##/BadgeTwinHandler
<4> 2021-05-12 13:23:57.896 +00:00 [WRN] - No subscription for receiving messages found for ##/BadgeTwinHandler
<4> 2021-05-12 13:23:57.945 +00:00 [WRN] - No subscription for receiving messages found for ##/ContactManager
<4> 2021-05-12 13:24:04.575 +00:00 [WRN] - No subscription for receiving messages found for ##/BadgeTwinHandler
<4> 2021-05-12 13:24:05.014 +00:00 [WRN] - No subscription for receiving messages found for ##/BadgeTwinHandler
<4> 2021-05-12 13:24:06.802 +00:00 [WRN] - No subscription for receiving messages found for ##/ContactManager
<4> 2021-05-12 13:24:19.061 +00:00 [WRN] - No subscription for receiving messages found for ##/BadgeTwinHandler
<4> 2021-05-12 13:24:22.342 +00:00 [WRN] - No subscription for receiving messages found for ##/ContactManager
<4> 2021-05-12 13:24:23.094 +00:00 [WRN] - No subscription for receiving messages found for ##/BadgeTwinHandler
<4> 2021-05-12 13:24:45.954 +00:00 [WRN] - No subscription for receiving messages found for ##/BadgeTwinHandler
<4> 2021-05-12 13:24:49.967 +00:00 [WRN] - No subscription for receiving messages found for ##/BadgeTwinHandler
<4> 2021-05-12 13:24:53.833 +00:00 [WRN] - No subscription for receiving messages found for ##/ContactManager
<4> 2021-05-12 13:25:45.383 +00:00 [WRN] - No subscription for receiving messages found for ##/ContactManager
<4> 2021-05-12 13:25:45.972 +00:00 [WRN] - No subscription for receiving messages found for ##/BadgeTwinHandler
<4> 2021-05-12 13:25:49.970 +00:00 [WRN] - No subscription for receiving messages found for ##/BadgeTwinHandler
<4> 2021-05-12 13:26:45.407 +00:00 [WRN] - No subscription for receiving messages found for ##/ContactManager
<4> 2021-05-12 13:26:45.982 +00:00 [WRN] - No subscription for receiving messages found for ##/BadgeTwinHandler
<4> 2021-05-12 13:26:49.983 +00:00 [WRN] - No subscription for receiving messages found for ##/BadgeTwinHandler
<4> 2021-05-12 13:27:45.414 +00:00 [WRN] - No subscription for receiving messages found for ##/ContactManager
<4> 2021-05-12 13:27:45.986 +00:00 [WRN] - No subscription for receiving messages found for ##/BadgeTwinHandler
<4> 2021-05-12 13:27:49.983 +00:00 [WRN] - No subscription for receiving messages found for ##/BadgeTwinHandler
<6> 2021-05-12 13:28:40.619 +00:00 [INF] - Entering periodic task to reauthenticate connected clients
```
